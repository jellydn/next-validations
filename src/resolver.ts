import { assert, type Infer, type Schema } from '@typeschema/main';

export function typeschemaResolver<T extends Schema>(schema: T) {
  return {
    async validate(data: unknown): Promise<Infer<T>> {
      return assert(schema, data);
    },
  };
}
