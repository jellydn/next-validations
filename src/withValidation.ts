import { type Schema } from '@typeschema/main';
import { type NextApiRequest, type NextApiResponse } from 'next';
import { typeschemaResolver } from './resolver';

type NextHandler = (err?: Error) => void;

type ValidationHoF = {
  mode?: 'body' | 'query' | 'headers';
  schema: Schema;
};

export function withValidation({ schema, mode = 'query' }: ValidationHoF) {
  return withValidations([{ schema, mode }]);
}

export function withValidations(validations: ValidationHoF[]) {
  return (handler?: (req: NextApiRequest, res: NextApiResponse) => any) =>
    async (req: NextApiRequest, res: NextApiResponse, next?: NextHandler) => {
      try {
        await Promise.all(
          validations.map(async (validation) => {
            const resolver = typeschemaResolver(validation.schema);
            await resolver.validate(req[validation.mode ?? 'query']);
          })
        );

        if (next) {
          next();
          return;
        }

        if (handler) return handler(req, res);

        res.status(404).end();
      } catch (error) {
        res.status(400).send(error);
      }
    };
}
