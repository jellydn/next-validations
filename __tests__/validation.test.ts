import Joi from 'joi';
import * as valibot from 'valibot';
import * as yup from 'yup';
import { z } from 'zod';
import { typeschemaResolver } from '../src/resolver';

describe('Yup Validation', () => {
  it('should return true', async () => {
    const schema = yup.object().shape({
      name: yup.string().required(),
      age: yup.number().required().positive().integer(),
      email: yup.string().email(),
      website: yup.string().url(),
      createdOn: yup.date().default(function () {
        return new Date();
      }),
    });
    const resolver = typeschemaResolver(schema);
    const isValid = await resolver.validate({
      name: 'jimmy',
      age: 24,
    });
    expect(isValid).toBeTruthy();
  });

  it('should return false', async () => {
    const schema = yup.object().shape({
      name: yup.string().required(),
      age: yup.number().required().positive().integer(),
      email: yup.string().email(),
      website: yup.string().url(),
      createdOn: yup.date().default(function () {
        return new Date();
      }),
    });
    const resolver = typeschemaResolver(schema);
    try {
      await resolver.validate({
        name: 'jimmy',
      });
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

describe('Yoi Validation', () => {
  it('should return true', async () => {
    const schema = Joi.object({
      dob: Joi.date().iso(),
      email: Joi.string().email().required(),
      name: Joi.string().required(),
    });

    const resolver = typeschemaResolver(schema);
    const isValid = await resolver.validate({
      name: 'Huynh Duc Dung',
      email: 'dung@productsway.com',
      dob: 1988,
    });
    expect(isValid).toBeTruthy();
  });

  it('should return false', async () => {
    const schema = Joi.object({
      dob: Joi.date().iso(),
      email: Joi.string().email().required(),
      name: Joi.string().required(),
    });
    const resolver = typeschemaResolver(schema);
    try {
      await resolver.validate({});
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

describe('Zod Validation', () => {
  it('should return true', async () => {
    const schema = z.object({
      username: z.string(),
    });

    const resolver = typeschemaResolver(schema);
    const isValid = await resolver.validate({
      username: 'jellydn',
    });
    expect(isValid).toBeTruthy();
  });

  it('should return false', async () => {
    const schema = z.object({
      username: z.string().min(8),
    });
    const resolver = typeschemaResolver(schema);
    try {
      await resolver.validate({
        username: 'jellydn',
      });
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

describe('Valibot Validation', () => {
  it('should return true', async () => {
    const schema = valibot.object({
      username: valibot.string(),
    });

    const resolver = typeschemaResolver(schema);
    const isValid = await resolver.validate({
      username: 'jellydn',
    });
    expect(isValid).toBeTruthy();
  });

  it('should return false', async () => {
    const schema = valibot.object({
      age: valibot.number(),
    });
    const resolver = typeschemaResolver(schema);
    try {
      await resolver.validate({
        age: '35',
      });
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});
