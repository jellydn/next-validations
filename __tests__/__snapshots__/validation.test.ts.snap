// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Valibot Validation > should return false 1`] = `[AggregateError: Assertion failed]`;

exports[`Yoi Validation > should return false 1`] = `[AggregateError: Assertion failed]`;

exports[`Yup Validation > should return false 1`] = `[AggregateError: Assertion failed]`;

exports[`Zod Validation > should return false 1`] = `[AggregateError: Assertion failed]`;
